geomstats/learning/kmeans.py

```diff
@@ -103,26 +108,50 @@ def fit(self, X):
         n_samples = X.shape[0]
         if self.verbose > 0:
             logging.info("Initializing...")
-        if self.init == "kmeans++":
-            centroids = [gs.expand_dims(X[randint(0, n_samples - 1)], 0)]
-            for i in range(self.n_clusters - 1):
-                dists = [
-                    gs.to_ndarray(self.metric.dist(centroids[j], X), 2, 1)
-                    for j in range(i + 1)
+
+        if isinstance(self.init, str):
+            if self.init == "kmeans++":
+                centroids = [gs.expand_dims(X[randint(0, n_samples - 1)], 0)]
+                for i in range(self.n_clusters - 1):
+                    dists = [
+                        gs.to_ndarray(self.metric.dist(centroids[j], X), 2, 1)
+                        for j in range(i + 1)
+                    ]
+                    dists = gs.hstack(dists)
+                    dists_to_closest_centroid = gs.amin(dists, 1)
+                    indices = gs.arange(n_samples)
+                    weights = dists_to_closest_centroid / gs.sum(
+                        dists_to_closest_centroid
+                    )
+                    index = rv_discrete(values=(indices, weights)).rvs()
+                    centroids.append(gs.expand_dims(X[index], 0))
+            elif self.init == "random":
+                centroids = [
+                    gs.expand_dims(X[randint(0, n_samples - 1)], 0)
+                    for i in range(self.n_clusters)
                 ]
-                dists = gs.hstack(dists)
-                dists_to_closest_centroid = gs.amin(dists, 1)
-                indices = gs.arange(n_samples)
-                weights = dists_to_closest_centroid / gs.sum(dists_to_closest_centroid)
-                index = rv_discrete(values=(indices, weights)).rvs()
-                centroids.append(gs.expand_dims(X[index], 0))
+            else:
+                raise ValueError(f"Unknown initial centroids method '{self.init}'.")
+
+            centroids = gs.concatenate(centroids, axis=0)
         else:
-            centroids = [
-                gs.expand_dims(X[randint(0, n_samples - 1)], 0)
-                for i in range(self.n_clusters)
-            ]
-        self.centroids = gs.concatenate(centroids, axis=0)
-        self.init_centroids = gs.concatenate(centroids, axis=0)
+            if callable(self.init):
+                centroids = self.init(X, self.n_clusters)
+            else:
+                centroids = self.init
+
+            if centroids.shape[0] != self.n_clusters:
+                raise ValueError("Need as many initial centroids as clusters.")
+
+            if centroids.shape[1] != X.shape[1]:
+                raise ValueError(
+                    "Dimensions of initial centroids and training data do not match."
+                )
+
+            centroids = gs.copy(centroids)
+
+        self.centroids = centroids
```

Comment:
Here we need to do:

```
self.centroids = gs.copy(centroids)
self.init_centroids = gs.copy(centroids)
```

otherwise both `self.centroids` and `self.init_centroids` will be the same object (at least in numpy the arrays are mutable).